File: src/Python/Linked-List/25.k-个一组翻转链表.py
#
# @lc app=leetcode.cn id=25 lang=python
#
# [25] K 个一组翻转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        pre = dummy
        end = dummy
        while end:
            for i in range(k):
                if end:
                    end = end.next
                else:
                    break
            if not end: break
            start = pre.next
            next = end.next
            end.next = None
            pre.next = self.reverse(start)
            start.next = next
            pre = start
            end = pre
        return dummy.next
    def reverse(self, head):
        pre = None
        cur = head
        while cur:
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next
        return pre
# @lc code=end

File: src/Javascript/Linked-List/25.k-个一组翻转链表.js
/*
 * @lc app=leetcode.cn id=25 lang=javascript
 *
 * [25] K 个一组翻转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    const dummy = new ListNode(0);
    dummy.next = head;
    let pre = dummy, end = dummy;
    while (end.next != null) {
        for (let i = 0; i < k && end != null; i++) end = end.next;
        // length lower than k
        if (end == null) break;
        // start of reverse
        let start = pre.next;
        // next node of end of reverse
        let next = end.next;
        // cut off lists
        end.next = null;
        // reverse k nodes
        pre.next = reverseList(start);
        start.next = next;
        pre = start;
        end = pre;
    }
    return dummy.next;
};
const reverseList = (head) => {
    let pre = null
    let curr = head
    while (curr != null) {
        let next = curr.next;
        curr.next = pre;
        pre = curr;
        curr = next;
    }
    return pre;
}
// @lc code=end

